cns/restserver/ipam.go

```diff
@@ -0,0 +1,290 @@
+	b, err := json.Marshal(orchestratorContext)
+	return &cns.ContainerIPConfigState{
+		IPConfig:            ipconfig,
+		ID:                  id,
+		NCID:                ncid,
+		State:               state,
+		OrchestratorContext: b,
+	}, err
+}
+
+// used to request an IPConfig from the CNS state
+func (service *HTTPRestService) requestIPConfigHandler(w http.ResponseWriter, r *http.Request) {
+	var (
+		err             error
+		ipconfigRequest cns.GetIPConfigRequest
+		ipState         *cns.ContainerIPConfigState
+		returnCode      int
+		returnMessage   string
+	)
+
+	err = service.Listener.Decode(w, r, &ipconfigRequest)
+	logger.Request(service.Name, &ipconfigRequest, err)
+	if err != nil {
+		return
+	}
+
+	// retrieve ipconfig from nc
+	if ipState, err = requestIPConfigHelper(service, ipconfigRequest); err != nil {
+		returnCode = UnexpectedError
+		returnMessage = fmt.Sprintf("AllocateIPConfig failed: %v", err)
+	}
+
+	resp := cns.Response{
+		ReturnCode: returnCode,
+		Message:    returnMessage,
+	}
+
+	reserveResp := &cns.GetIPConfigResponse{
+		Response: resp,
+	}
+	reserveResp.IPConfiguration.IPSubnet = ipState.IPConfig
+
+	err = service.Listener.Encode(w, &reserveResp)
+	logger.Response(service.Name, reserveResp, resp.ReturnCode, ReturnCodeToString(resp.ReturnCode), err)
+}
+
+func (service *HTTPRestService) releaseIPConfigHandler(w http.ResponseWriter, r *http.Request) {
+	var (
+		podInfo    cns.KubernetesPodInfo
+		req        cns.GetIPConfigRequest
+		statusCode int
+	)
+
+	statusCode = UnexpectedError
+
+	err := service.Listener.Decode(w, r, &req)
+	logger.Request(service.Name, &req, err)
+	if err != nil {
+		return
+	}
+
+	defer func() {
+		resp := cns.Response{}
+
+		if err != nil {
+			resp.ReturnCode = statusCode
+			resp.Message = err.Error()
+		}
+
+		err = service.Listener.Encode(w, &resp)
+		logger.Response(service.Name, resp, resp.ReturnCode, ReturnCodeToString(resp.ReturnCode), err)
+	}()
+
+	if service.state.OrchestratorType != cns.Kubernetes {
+		err = fmt.Errorf("ReleaseIPConfig API supported only for kubernetes orchestrator")
+		return
+	}
+
+	// retrieve podinfo  from orchestrator context
+	if err = json.Unmarshal(req.OrchestratorContext, &podInfo); err != nil {
+		return
+	}
+
+	if err = service.ReleaseIPConfig(podInfo); err != nil {
+		statusCode = NotFound
+		return
+	}
+	return
+}
+
+//AddIPConfigsToState takes a lock on the service object, and will add an array of ipconfigs to the CNS Service.
+//Used to add IPConfigs to the CNS pool, specifically in the scenario of rebatching.
+func (service *HTTPRestService) AddIPConfigsToState(ipconfigs []*cns.ContainerIPConfigState) error {
+	service.Lock()
+
+	for i, ipconfig := range ipconfigs {
+		service.PodIPConfigState[ipconfig.ID] = ipconfig
+
+		if ipconfig.State == cns.Allocated {
+			var podInfo cns.KubernetesPodInfo
+
+			if err := json.Unmarshal(ipconfig.OrchestratorContext, &podInfo); err != nil {
+				service.Unlock()
```

Comment:
can you add defer service.Unlock() instead of calling service.Unlock at multiple places

Reply:
updated